device/pci_device.cpp

```diff
@@ -0,0 +1,431 @@
+/*
+ * SPDX-FileCopyrightText: (c) 2023 Tenstorrent Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <cstdint>
+#include <vector>
+#include <fcntl.h>  // for ::open
+#include <unistd.h> // for ::close
+#include <sys/ioctl.h> // for ioctl
+#include <sys/mman.h>  // for mmap, munmap
+#include <linux/pci.h> // for PCI_SLOT, PCI_FUNC
+
+#include "pci_device.hpp"
+#include "architecture_implementation.h"
+#include "ioctl.h"
+#include "device/tt_arch_types.h"
+#include "device/driver_atomics.h"
+
+#include "common/assert.hpp"
+#include "common/logger.hpp"
+
+int find_device(const uint16_t device_id) {
+    // returns device id if found, otherwise -1
+    const char device_name_pattern [] = "/dev/tenstorrent/%u";
+    char device_name[sizeof(device_name_pattern) + std::numeric_limits<unsigned int>::digits10];
+    std::snprintf(device_name, sizeof(device_name), device_name_pattern, (unsigned int)device_id);
```

Comment:
A small cleanup could be to use fmt for string formatting everywhere (it's already a submodule)